File: src/main/java/me/pcasaes/hexoids/infrastructure/producer/DomainEventProducer.java
package me.pcasaes.hexoids.infrastructure.producer;

import me.pcasaes.hexoids.core.domain.model.DomainEvent;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import pcasaes.hexoids.proto.Event;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import java.time.Duration;
import java.util.Properties;
import java.util.UUID;

/**
 * Used to generate domain events. Domain events are used to keep server nodes in sync
 */
@ApplicationScoped
public class DomainEventProducer {

    private static final String NAME = "domain-event-producer";


    private final String kafkaBroker;

    private Producer<UUID, Event> producer;

    @Inject

    public DomainEventProducer(
            @ConfigProperty(name = "kafka.bootstrap.servers") String kafkaBroker
    ) {
        this.kafkaBroker = kafkaBroker;
    }

    @PostConstruct
    public void start() {
        this.producer = createProducer();
    }

    @PreDestroy
    public void stop() {
        this.producer.close(Duration.ofSeconds(5));
    }

    public void accept(DomainEvent event) {
        if (event != null) {
            this.producer.send(new ProducerRecord<>(event.getTopic(), event.getKey(), event.getEvent()));
        }
    }

    public String getName() {
        return NAME;
    }

    public Producer<UUID, Event> createProducer() {

        Properties props = new Properties();

        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, this.kafkaBroker);
        props.put(ProducerConfig.CLIENT_ID_CONFIG, UUID.randomUUID().toString());

        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "me.pcasaes.hexoids.infrastructure.kafka.converter.UUIDBytesSerializer");
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "me.pcasaes.hexoids.infrastructure.kafka.converter.EventDtoSerializer");
        props.put(ProducerConfig.ACKS_CONFIG, "0");

        return new KafkaProducer<>(props);

    }
}

File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/producer/DomainEventProducer.kt
package me.pcasaes.hexoids.infrastructure.producer

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import me.pcasaes.hexoids.core.domain.model.DomainEvent
import me.pcasaes.hexoids.core.domain.model.GameTopic
import me.pcasaes.hexoids.core.domain.model.EventRecord
import org.eclipse.microprofile.reactive.messaging.Channel
import org.eclipse.microprofile.reactive.messaging.Emitter
import org.eclipse.microprofile.reactive.messaging.OnOverflow

/**
 * Used to generate domain events. Domain events are used to keep server nodes in sync
 */
@ApplicationScoped
class DomainEventProducer @Inject constructor(
    @Channel("join-game-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) joinGameEmitter: Emitter<EventRecord>,
    @Channel("player-action-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) playerActionEmitter: Emitter<EventRecord>,
    @Channel("bolt-life-cycle-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) boltLifeCycleEmitter: Emitter<EventRecord>,
    @Channel("bolt-action-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) boltActionEmitter: Emitter<EventRecord>,
    @Channel("score-board-control-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) scoreBoardControlEmitter: Emitter<EventRecord>,
    @Channel("score-board-update-out") @OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) scoreBoardUpdateEmitter: Emitter<EventRecord>
) {
    private val emitters: Array<Emitter<EventRecord>?>

    init {
        val em = arrayOfNulls<Emitter<EventRecord>?>(GameTopic.entries.size)
        em[GameTopic.JOIN_GAME_TOPIC.ordinal] = joinGameEmitter
        em[GameTopic.PLAYER_ACTION_TOPIC.ordinal] = playerActionEmitter
        em[GameTopic.BOLT_LIFECYCLE_TOPIC.ordinal] = boltLifeCycleEmitter
        em[GameTopic.BOLT_ACTION_TOPIC.ordinal] = boltActionEmitter
        em[GameTopic.SCORE_BOARD_CONTROL_TOPIC.ordinal] = scoreBoardControlEmitter
        em[GameTopic.SCORE_BOARD_UPDATE_TOPIC.ordinal] = scoreBoardUpdateEmitter

        this.emitters = em
    }

    fun accept(event: DomainEvent?) {
        val topic = event?.topic
        if (topic != null) {
            this.emitters[GameTopic.valueOf(topic).ordinal]
                ?.send(EventRecord.of(event))
        }
    }
}
